File: Stripe/STPTelemetryClient.m
//
//  STPTelemetryClient.m
//  Stripe
//
//  Created by Ben Guo on 4/18/17.
//  Copyright © 2016 Stripe, Inc. All rights reserved.
//

#import <UIKit/UIKit.h>
#import <sys/utsname.h>

#import "NSBundle+Stripe_AppName.h"
#import "STPTelemetryClient.h"
#import "STPAPIClient.h"
#import "STPAPIClient+Private.h"

@interface STPTelemetryClient ()
@property (nonatomic) NSDate *appOpenTime;
@property (nonatomic, readwrite) NSURLSession *urlSession;
@end

@implementation STPTelemetryClient

+ (BOOL)shouldSendTelemetry {
#if TARGET_OS_SIMULATOR
    return NO;
#else
    return NSClassFromString(@"XCTest") == nil;
#endif
}

+ (instancetype)sharedInstance {
    NSURLSessionConfiguration *config = [STPAPIClient sharedUrlSessionConfiguration];
    static STPTelemetryClient *sharedClient;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        sharedClient = [[self alloc] initWithSessionConfiguration:config];
    });
    return sharedClient;
}

- (instancetype)init {
    return [self initWithSessionConfiguration:[STPAPIClient sharedUrlSessionConfiguration]];
}

- (instancetype)initWithSessionConfiguration:(NSURLSessionConfiguration *)config {
    self = [super init];
    if (self) {
        _urlSession = [NSURLSession sessionWithConfiguration:config];
        [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(applicationDidBecomeActive) name:UIApplicationDidBecomeActiveNotification object:nil];
        [[UIDevice currentDevice] setBatteryMonitoringEnabled:YES];
    }
    return self;
}

- (void)dealloc {
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

- (void)addTelemetryFieldsToParams:(NSMutableDictionary *)params {
    params[@"muid"] = [self muid];
    params[@"time_on_page"] = [self timeOnPage];
}

- (void)applicationDidBecomeActive {
    self.appOpenTime = [NSDate date];
}

- (NSString *)muid {
    NSString *muid = [[[UIDevice currentDevice] identifierForVendor] UUIDString];
    return muid ?: @"";
}

- (NSNumber *)timeOnPage {
    if (!self.appOpenTime) {
        return @(0);
    }
    NSTimeInterval seconds = [[NSDate date] timeIntervalSinceDate:self.appOpenTime];
    NSInteger millis = (NSInteger)round(seconds*1000);
    return @(MAX(millis, 0));
}

- (NSString *)language {
    NSString *localeID = [[NSLocale currentLocale] localeIdentifier];
    return localeID ?: @"";
}

- (NSString *)platform {
    return [@[[self deviceModel], [self osVersion]] componentsJoinedByString:@" "];
}

- (NSString *)deviceModel {
    struct utsname systemInfo;
    uname(&systemInfo);
    return @(systemInfo.machine) ?: @"";
}

- (NSString *)osVersion {
    return [UIDevice currentDevice].systemVersion ?: @"";
}

- (NSString *)screenSize {
    UIScreen *screen = [UIScreen mainScreen];
    CGRect screenRect = [screen bounds];
    CGFloat width = screenRect.size.width;
    CGFloat height = screenRect.size.height;
    CGFloat scale = [screen scale];
    return [NSString stringWithFormat:@"%.0fw_%.0fh_%.0fr", width, height, scale];
}

- (NSString *)timeZoneOffset {
    NSTimeZone *timeZone = [NSTimeZone localTimeZone];
    double hoursFromGMT = (double)timeZone.secondsFromGMT/(60*60);
    return [NSString stringWithFormat:@"%.0f", hoursFromGMT];
}

- (NSDictionary *)encodeValue:(NSString *)value {
    if (value) {
        return @{@"v": value};
    }
    return nil;
}

- (NSDictionary *)payload {
    NSMutableDictionary *payload = [NSMutableDictionary new];
    NSMutableDictionary *data = [NSMutableDictionary new];
    data[@"c"] = [self encodeValue:[self language]];
    data[@"d"] = [self encodeValue:[self platform]];
    data[@"f"] = [self encodeValue:[self screenSize]];
    data[@"g"] = [self encodeValue:[self timeZoneOffset]];
    payload[@"a"] = [data copy];
    NSMutableDictionary *otherData = [NSMutableDictionary new];
    otherData[@"d"] = [self muid];
    otherData[@"k"] = [NSBundle stp_applicationName];
    otherData[@"l"] = [NSBundle stp_applicationVersion];
    otherData[@"m"] = @([Stripe deviceSupportsApplePay]);
    otherData[@"o"] = [self osVersion];
    otherData[@"s"] = [self deviceModel];
    payload[@"b"] = [otherData copy];
    payload[@"tag"] = STPSDKVersion;
    payload[@"src"] = @"ios-sdk";
    payload[@"v2"] = @1;
    return [payload copy];
}

- (void)sendTelemetryData {
    if (![[self class] shouldSendTelemetry]) {
        return;
    }
    NSString *path = @"ios-sdk-1";
    NSURL *url = [[NSURL URLWithString:@"https://m.stripe.com"] URLByAppendingPathComponent:path];
    NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:url];
    request.HTTPMethod = @"POST";
    [request setValue:@"application/json" forHTTPHeaderField:@"Content-Type"];
    NSDictionary *payload = [self payload];
    NSData *data = [NSJSONSerialization dataWithJSONObject:payload options:(NSJSONWritingOptions)0 error:nil];
    request.HTTPBody = data;
    NSURLSessionDataTask *task = [self.urlSession dataTaskWithRequest:request];
    [task resume];
}

@end

File: Stripe/STPTelemetryClient.swift
//
//  STPTelemetryClient.swift
//  Stripe
//
//  Created by Ben Guo on 4/18/17.
//  Copyright © 2016 Stripe, Inc. All rights reserved.
//

import Foundation
import UIKit

final class STPTelemetryClient: NSObject {
  @objc(sharedInstance) static var shared: STPTelemetryClient = STPTelemetryClient(sessionConfiguration: STPAPIClient.sharedUrlSessionConfiguration)

  func addTelemetryFields(toParams params: inout [String: Any]) {
    params["muid"] = muid
  }

  @objc func paramsByAddingTelemetryFields(toParams params: [String: Any]) -> [String: Any] {
    var mutableParams = params
    mutableParams["muid"] = muid
    return mutableParams
  }

  func sendTelemetryData() {
    guard STPTelemetryClient.shouldSendTelemetry() else {
      return
    }
    let path = "ios-sdk-1"
    let url = URL(string: "https://m.stripe.com")!.appendingPathComponent(path)
    let request = NSMutableURLRequest(url: url)
    request.httpMethod = "POST"
    request.setValue("application/json", forHTTPHeaderField: "Content-Type")
    let payload = self.payload
    var data: Data?
    do {
      data = try JSONSerialization.data(
        withJSONObject: payload, options: JSONSerialization.WritingOptions(rawValue: 0))
    } catch {
      return
    }
    request.httpBody = data
    let task = urlSession.dataTask(with: request as URLRequest)
    task.resume()
  }

  private let urlSession: URLSession

  class func shouldSendTelemetry() -> Bool {
    #if targetEnvironment(simulator)
      return false
    #else
      return StripeAPI.advancedFraudSignalsEnabled && NSClassFromString("XCTest") == nil
    #endif
  }

  init(sessionConfiguration config: URLSessionConfiguration) {
    urlSession = URLSession(configuration: config)
    super.init()
  }

  private var muid : String {
    get {
      let muid = UIDevice.current.identifierForVendor?.uuidString
      return muid ?? ""
    }
  }

  private var language = Locale.autoupdatingCurrent.identifier

  lazy private var platform = [deviceModel, osVersion].joined(separator: " ")

  private var deviceModel : String = {
    var systemInfo = utsname()
    uname(&systemInfo)
    let model = withUnsafePointer(to: &systemInfo.machine) {
      $0.withMemoryRebound(to: CChar.self, capacity: 1) {
        ptr in String.init(validatingUTF8: ptr)
      }
    }
    return model ?? "Unknown"
  }()

  private var osVersion = UIDevice.current.systemVersion

  private var screenSize : String {
    let screen = UIScreen.main
    let screenRect = screen.bounds
    let width = screenRect.size.width
    let height = screenRect.size.height
    let scale = screen.scale
    return String(format: "%.0fw_%.0fh_%.0fr", width, height, scale)
  }

  private var timeZoneOffset : String {
    let timeZone = NSTimeZone.local as NSTimeZone
    let hoursFromGMT = Double(timeZone.secondsFromGMT) / (60 * 60)
    return String(format: "%.0f", hoursFromGMT)
  }

  private func encodeValue(_ value: String?) -> [AnyHashable: Any]? {
    if let value = value {
      return [
        "v": value
      ]
    }
    return nil
  }

  private var payload : [AnyHashable: Any] {
    var payload: [AnyHashable: Any] = [:]
    var data: [AnyHashable: Any] = [:]
    if let encode = encodeValue(language) {
      data["c"] = encode
    }
    if let encode = encodeValue(platform) {
      data["d"] = encode
    }
    if let encode = encodeValue(screenSize) {
      data["f"] = encode
    }
    if let encode = encodeValue(timeZoneOffset) {
      data["g"] = encode
    }
    payload["a"] = data
    var otherData: [AnyHashable: Any] = [:]
    otherData["d"] = muid
    otherData["k"] = Bundle.stp_applicationName() ?? ""
    otherData["l"] = Bundle.stp_applicationVersion() ?? ""
    otherData["m"] = NSNumber(value: StripeAPI.deviceSupportsApplePay())
    otherData["o"] = osVersion
    otherData["s"] = deviceModel
    payload["b"] = otherData
    payload["tag"] = STPAPIClient.STPSDKVersion
    payload["src"] = "ios-sdk"
    payload["v2"] = NSNumber(value: 1)
    return payload
  }
}
